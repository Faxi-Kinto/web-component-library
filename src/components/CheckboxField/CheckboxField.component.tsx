import React, { useMemo } from 'react';
import * as Styled from './CheckboxField.styles';
import { FieldProps } from '../../helpers/FieldProps';
import useField from '../../hooks/useField';
import { CheckboxProps } from '../Checkbox/Checkbox.component';
import Text from '../Text';
import Checkbox from '../Checkbox';

/**
 * @name CheckboxField
 *
 * [Insert omponent description]
 *
 * @returns {JSX}
 */

export type CheckboxFieldProps = FieldProps<boolean> & {
  description?: React.ReactNode;
} & CheckboxProps;

const CheckboxField: React.FC<CheckboxFieldProps> = (
  props: CheckboxFieldProps
): JSX.Element => {
  const { value, onChange, description, error, ...rest } = props;

  const showError = useField(props);

  const descriptION = useMemo(() => {
    if (!description) return null;

    if (typeof description === 'string') {
      return (
        <Text.Body className="checkbox-field__description">
          {description}
        </Text.Body>
      );
    }
    return description;
  }, [description]);

  return (
    <Styled.Container>
      <Checkbox
        {...rest}
        checked={value}
        onChange={onChange ? ev => onChange(ev.target.checked) : undefined}
        error={showError}
      />
      {descriptION}
      {showError ? (
        <div className="checkbox-field__error-msg">{error}</div>
      ) : (
        ''
      )}
    </Styled.Container>
  );
};

CheckboxField.defaultProps = {};

export default CheckboxField;
