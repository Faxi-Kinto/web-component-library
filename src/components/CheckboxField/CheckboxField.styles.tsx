import styled from 'styled-components';
import { pxToRem } from '../../styles/basics';
import { theme } from '../../config';

export const Container = styled.div`
  line-height: ${pxToRem('20px')};

  .checkbox-field {
    &__description,
    &__error-msg {
      margin-top: ${pxToRem('20px')};
      padding-right: ${pxToRem('50px')};
      font-size: ${pxToRem('14px')};

      &__description {
        display: block;
      }

      &__error-msg {
        color: ${theme.pallet.ERROR};
      }
    }
  }
`;
