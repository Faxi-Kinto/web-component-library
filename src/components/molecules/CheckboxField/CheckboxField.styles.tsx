import styled from 'styled-components';
import { pxToRem } from 'styles/fonts';
import { theme } from 'config';

export const Container = styled.div`
  line-height: ${pxToRem('20px')};

  .checkbox-field__description,
  .checkbox-field__error-msg {
    margin-top: 20px;
    padding-right: 50px;
    font-size: ${pxToRem('14px')};
  }
  .checkbox-field__description {
    display: block;
  }
  .checkbox-field__error-msg {
    color: ${theme.pallet.ERROR};
  }
`;
