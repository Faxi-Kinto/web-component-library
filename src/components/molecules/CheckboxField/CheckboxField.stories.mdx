import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import { withKnobs, text, boolean } from '@storybook/addon-knobs';

import CheckboxField from './';

<Meta title="Core|Molecules/CheckboxField" />

# CheckboxField Component

This field is a wrapper around [Checkbox](/docs/core-atoms-checkbox--default-story) atom and it is intended for use in [Form Field](/docs/core-molecules-formfield--default-story)
enabeling error messages, error state for input atom and description message for the field.
`CheckboxField` implements FormField API.

#### Default

<Preview>
  <Story name="default" decorators={[withKnobs]}>
    <CheckboxField
      label="Click me!"
      error={text('Error message', 'This field is required')}
      description={text('Description message', '')}
      labelPosition={(boolean('Label on left',false)) ? 'left':'right'}
      touched={boolean('Touched', false)}
      dirty={boolean('Dirty', false)}
    />
  </Story>
</Preview>

#### With Error message

<Preview>
  <Story name="With error message">
    <CheckboxField
      label="Click me!"
      error="Error message"
      touched={true}
      dirty={true}
    />
  </Story>
</Preview>

#### With description message

<Preview>
  <Story name="With description message">
    <CheckboxField
      label="Click me!"
      description="This is a description message for this checkbox"
    />
  </Story>
</Preview>

#### With position left description message

<Preview>
  <Story name="With position left description message">
    <CheckboxField
      label="Click me!"
      labelPosition={'left'}
      description="Faxi would like to contact you via in-app messages, push messages and email to promote other Faxi services and the services of selected third parties which we believe may be of interest to you. Please see our Privacy Policy for more details."
    />
  </Story>
</Preview>

## Props

<Props of={CheckboxField} />
