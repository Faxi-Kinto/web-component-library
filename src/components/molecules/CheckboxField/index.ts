import CheckboxField from './CheckboxField.component';

export default CheckboxField;
