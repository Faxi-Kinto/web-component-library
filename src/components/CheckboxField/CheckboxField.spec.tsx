import React from 'react';
import renderer from 'react-test-renderer';

import CheckboxField from './CheckboxField.component';

test('CheckboxField renders correctly', () => {
  const component = renderer.create(<CheckboxField />);

  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
