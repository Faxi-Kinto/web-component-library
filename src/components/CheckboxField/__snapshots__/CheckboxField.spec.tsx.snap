// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxField renders correctly 1`] = `
.c0 {
  line-height: 1.25rem;
}

.c0 .checkbox-field__description,
.c0 .checkbox-field__error-msg {
  margin-top: 1.25rem;
  padding-right: 3.125rem;
  font-size: 0.875rem;
}

.c0 .checkbox-field__description__description,
.c0 .checkbox-field__error-msg__description {
  display: block;
}

.c0 .checkbox-field__description__error-msg,
.c0 .checkbox-field__error-msg__error-msg {
  color: #EB4820;
}

.c1 {
  position: relative;
  cursor: pointer;
  min-height: 1.25rem;
  padding-left: 1.25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: initial;
  -webkit-justify-content: initial;
  -ms-flex-pack: initial;
  justify-content: initial;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 > input {
  display: none;
}

.c1:before {
  content: '';
  display: inline-block;
  box-sizing: border-box;
  border: #4E606A solid 0.0625rem;
  position: absolute;
  top: 0;
  left: 0;
  width: 1.25rem;
  height: 1.25rem;
}

.c1 > span {
  line-height: 1.25rem;
}

.c1.checked:after {
  background: url('assets/icons/checkmark.svg') 100% no-repeat;
  content: '';
  width: 0.8125rem;
  height: 0.5625rem;
  position: absolute;
  top: 0.34375rem;
  left: 0.1875rem;
}

.c1.label--has-label-text {
  padding-left: 1.75rem;
}

.c1.label--has-error > input {
  color: #EB4820;
}

.c1.label--has-error:before {
  border: #EB4820 solid 0.0625rem;
}

.c1.label--left {
  padding: 0 1.75rem 0 0;
}

.c1.label--left:before {
  position: absolute;
  top: 0;
  right: 0;
  left: auto;
}

.c1.label--left.checked:after {
  position: absolute;
  top: 0.34375rem;
  right: 0.25rem;
  left: auto;
}

<div
  className="c0"
>
  <label
    className="c1"
  >
    <input
      checked={false}
      onChange={[Function]}
      type="checkbox"
    />
  </label>
  
</div>
`;
